lib/nandi/multi_database.rb

```diff
@@ -0,0 +1,157 @@
+# frozen_string_literal: true
+
+module Nandi
+  class MultiDatabase
+
+    class Database
+      # Most DDL changes take a very strict lock, but execute very quickly. For these
+      # the statement timeout should be very tight, so that if there's an unexpected
+      # delay the query queue does not back up.
+      DEFAULT_ACCESS_EXCLUSIVE_STATEMENT_TIMEOUT = 1_500
+      DEFAULT_ACCESS_EXCLUSIVE_LOCK_TIMEOUT = 5_000
+
+      DEFAULT_ACCESS_EXCLUSIVE_STATEMENT_TIMEOUT_LIMIT =
+        DEFAULT_ACCESS_EXCLUSIVE_STATEMENT_TIMEOUT
+      DEFAULT_ACCESS_EXCLUSIVE_LOCK_TIMEOUT_LIMIT =
+        DEFAULT_ACCESS_EXCLUSIVE_LOCK_TIMEOUT
+      DEFAULT_CONCURRENT_TIMEOUT_LIMIT = 3_600_000
+
+      DEFAULT_MIGRATION_DIRECTORY = "db/safe_migrations"
+      DEFAULT_OUTPUT_DIRECTORY = "db/migrate"
+
+      # The default lock timeout for migrations that take ACCESS EXCLUSIVE
+      # locks. Can be overridden by way of the `set_lock_timeout` class
+      # method in a given migration. Default: 1500ms.
+      # @return [Integer]
+      attr_accessor :access_exclusive_lock_timeout
+
+      # The default statement timeout for migrations that take ACCESS EXCLUSIVE
+      # locks. Can be overridden by way of the `set_statement_timeout` class
+      # method in a given migration. Default: 1500ms.
+      # @return [Integer]
+      attr_accessor :access_exclusive_statement_timeout
+
+      # The maximum lock timeout for migrations that take an ACCESS EXCLUSIVE
+      # lock and therefore block all reads and writes. Default: 5,000ms.
+      # @return [Integer]
+      attr_accessor :access_exclusive_statement_timeout_limit
+
+      # The maximum statement timeout for migrations that take an ACCESS
+      # EXCLUSIVE lock and therefore block all reads and writes. Default: 1500ms.
+      # @return [Integer]
+      attr_accessor :access_exclusive_lock_timeout_limit
+
+      # The minimum statement timeout for migrations that take place concurrently.
+      # Default: 3,600,000ms (ie, 3 hours).
+      # @return [Integer]
+      attr_accessor :concurrent_statement_timeout_limit
+
+      # The minimum lock timeout for migrations that take place concurrently.
+      # Default: 3,600,000ms (ie, 3 hours).
+      # @return [Integer]
+      attr_accessor :concurrent_lock_timeout_limit
+
+      # The directory for output files. Default: `db/migrate`
+      # @return [String]
+      attr_accessor :output_directory
+
+      attr_reader :name, :default
+
+      attr_accessor :migration_directory,
+                    :lockfile_name
+
+      def initialize(name:, config:)
```

Comment:
```suggestion
      def initialize(name: nil, config:)
```
Name can be nil for single db configs (Names only exist in the multi-db mode). We should probably add some validation that name is only nil when `default: true` and `databases.length == 1`

Reply:
actually, maybe best we add types now so we don't need to handle it